### PR TITLE
Unify dev-tools runners and convert to Typescript

### PR DESCRIPTION
**Pull request by scheibo**
- Refactor dev-tools/harness.js and dev-tools/smokes.js to separate out the script/CLI code and implementation code into separate files.
- Rename 'smoke' to 'exhaustive' ('multi' mode can also be used for "smoke testing") to better describe its behavior.
- Rewrite the runners in Typescript for type safety.
- Refactor common build utilities into dev-tools/build.js and introduce the notion of a 'full' build analogous to 'full' tests.

### CODE REVIEW

**Comment by Zarel** [comment]
This pull request **introduces 2 alerts** when merging 7f0c5f684bb7572fc141c7186278608686168250 into eb6c4384664de2aab1a905786e8793c902f37fe3 - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-043258aac482914a689d3b3db771288c888302c8)

**new alerts:**

* 2 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by scheibo** [comment]
LGTM is not on Typescript 3.4 yet, so it reporting false positives.

**Comment by Zarel** [comment]
This pull request **introduces 2 alerts** when merging 7e70a3db020158dd7988cc01fea65b1f63b727a3 into 3e735aa6ce89d88d4c1288cfe95aa055aaf24e0a - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-bc2449d9bc887b982ca70422f5396e0aa1e03bea)

**new alerts:**

* 2 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Zarel** [comment]
Can you hide instead of delete?

File: tslint.json [hunk 1]
```diff
@@ -68,6 +68,6 @@
     },
     "rulesDirectory": [],
     "linterOptions": {
-      "exclude": ["node_modules/**/*", "**/*.js", "dev-tools/*", ".*-dist/**/*"]
+      "exclude": ["node_modules/**/*", "**/*.js", "dev-tools/global*", ".*-dist/**/*"]
```
Comment (Zarel) [inline tslint.json]: Why isn't this just `"node_modules/"`?

Comment (Zarel) [reply tslint.json]: `**/*.js` is also a pretty inefficient glob... why would tslint bother with JS files at all?

**Comment by Zarel** [comment]
Maybe `runners.ts` should be moved into `sim/`?

**Comment by Zarel** [comment]
I'm also not sold on `runners` as a name.

**Comment by Zarel** [comment]
Maybe something like `automator`

**Comment by scheibo** [comment]
> Maybe `runners.ts` should be moved into `sim/`?

I think we need a separate tree for this + `random-player-ai.ts`. I feel depending on `random-player-ai.ts` in `sim/examples` is kind of 'eh', but I don't think either belong in `sim` proper.

> I'm also not sold on `runners` as a name.

Yeah its pretty terrible, I'll do some thinking. I'm not sold on `automator`, but thanks for at least suggesting an alternative.

> Can you hide comments instead of deleting them?

Oh true, didn't realize it wouldn't let other collaborators peek into them if they cared. Will do going forward

**Comment by Zarel** [comment]
I wouldn't mind merging as-is, but not having to build `sim` and `dev-tools` separately would be a nice thing to have before merging, as would a saner `tslint.json`.

**Comment by scheibo** [comment]
I'm in no hurry to merge, this is just a clean up anyway. I'll look into both of those.

File: tslint.json (hunk 1 above)
Comment (scheibo) [reply tslint.json]: > Why isn't this just `"node_modules/"`?

Done.

> **/*.js is also a pretty inefficient glob... why would tslint bother with JS files at all?

Good question. I've removed `jsRules`, but even without that it still checks JS! Maybe has something to do with our `checkJs`/`allowJs` settings?. Seems borked, but if I remove the JS glob from exclude i get zillions of complaints about our JS files.

**Comment by scheibo** [comment]
As discussed on chat, will merge and followup with a move into a `sim` subdirectory + naming clean up as part of an effort to nuke `dev-tools` as opposed to keeping this large refactor PR open too long.

**Comment by Zarel** [comment]
This pull request **introduces 2 alerts** when merging 915b999ea3ad2692c71537d94a2af3c1bec8053d into 5c10ebf42868901a3a7ec14cb9b777707e6bcec9 - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-a7dfed66044f94e7af3df0a54bb33d121974a093)

**new alerts:**

* 2 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Zarel** [comment]
This pull request **introduces 2 alerts** when merging 7d0a6e23eea4be588b16fda38ef9a776e008425f into cd82475d068bde05ad687fc666c42ca642bd3d18 - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-743c4173fd458c42efc63d79dc8242cf5cce5deb)

**new alerts:**

* 2 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**